src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -2050,6 +2078,82 @@ def _process_derived_type_decln(self, parent, decl, visibility_map):
 
+    def _process_derived_type_contains_block(self, parent, decl, tsymbol):
+        '''
+        Process the supplied fparser2 parse tree for a the CONTAINS section of
+        a derived-type declaration to add any procedures to the DataTypeSymbol.
+
+        :param parent: PSyIR node in which to insert the symbols found.
+        :type parent: :py:class:`psyclone.psyir.nodes.ScopingNode`
+        :param decl: fparser2 parse tree of declaration to process.
+        :type decl: :py:class:`fparser.two.Fortran2003.Type_Declaration_Stmt`
+        :param tsymbol: the DataTypeSymbol representing the derived-type.
+        :type tsymbol: :py:class:`psyclone.psyir.symbols.DataTypeSymbol`
+
+        :returns: None
+
+        '''
+
+        contains_blocks = walk(decl, Fortran2003.Type_Bound_Procedure_Part)
+        if contains_blocks:
+            # Get it.
+            contains = contains_blocks[0]
+            # Get all procedures in the CONTAINS section.
+            procedures = walk(contains, Fortran2003.Specific_Binding)
+
+            # Process each procedure.
+            for procedure in procedures:
+                supported = True
+                # We do not support interfaces.
+                if procedure.items[0] is not None:
+                    supported = False
+                # We do not support 'pass', 'nopass', 'deferred', etc.
+                if procedure.items[1] is not None:
+                    supported = False
+
+                # Get the name, look it up in the symbol table and
+                # get its datatype or create it if it does not exist.
+                procedure_name = procedure.items[3].string
+                procedure_symbol = parent.symbol_table.lookup(
+                    procedure_name,
+                    otherwise=None)
+                if procedure_symbol and supported:
+                    procedure_datatype = procedure_symbol.datatype
+                else:
+                    procedure_datatype = UnsupportedFortranType(
+                                            procedure.string,
+                                            None)
+
+                # Get the visibility of the procedure.
+                procedure_vis = tsymbol.visibility
+                if procedure.items[1] is not None:
+                    access_spec = walk(procedure.items[1],
+                                       Fortran2003.Access_Spec)
+                    if access_spec:
+                        procedure_vis = _process_access_spec(
+                            access_spec[0])
+
+                # Deal with the optional initial value.
+                # This could be, e.g., `null` in `procedure, name => null()` or
+                # `testkern_code` in `procedure :: code => testkern_code`
+                if procedure.items[4] is not None:
+                    initial_value_name = procedure.items[4].string
+                    initial_value_symbol = parent.symbol_table.lookup(
```

Comment:
I think `find_or_create` (https://psyclone-ref.readthedocs.io/en/latest/_static/html/classpsyclone_1_1psyir_1_1symbols_1_1symbol__table_1_1SymbolTable.html#ada706dd884505330c3d24b9b9a738042) would simplify things here. You can pass the `symboltype` and `data_type` that you want (if it doesn't exist) as arguments.

Reply:
Done, it's indeed clearer.